clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```
@@ -147,6 +151,10 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     // to keep from repeatedly scanning subscriptions in poll(), cache the result during metadata updates
     private boolean cachedSubscriptionHasAllFetchPositions;
     private final WakeupTrigger wakeupTrigger = new WakeupTrigger();
+    private volatile boolean isFenced = false;
```

I'm wondering why we need this member. It looks like the volatile boolean is written and read only from the application thread. Could it then just be a boolean? But if so, could we just throw the exception directly and avoid the indirection?

---

Thanks, we don't need this to be volatile.  In terms of exception throwing, I think once the instance is fenced, the we want to keep throwing if the user tries to commit or poll, because the expectation is that the user should close and reconfigure the instance.  See the original implementation

```
void invokeCompletedOffsetCommitCallbacks() {
        if (asyncCommitFenced.get()) {
            throw new FencedInstanceIdException("Get fenced exception for group.instance.id "
                + rebalanceConfig.groupInstanceId.orElse("unset_instance_id")
                + ", current member.id is " + memberId());
        }
         ...
}
```

Here asyncCommitFenced is set during the callback failure in the original implementation.
```
if (commitException instanceof FencedInstanceIdException) {
                    asyncCommitFenced.set(true);
}
```

---

Make sense, thanks!

---

@philipnee Can we remove the volatile if we do not need this?